contracts/InterestRateModels/TwoKinksInterestRateModel.sol

```diff
@@ -0,0 +1,178 @@
+        BASE_RATE_2_PER_BLOCK = baseRate2PerYear_ / BLOCKS_PER_YEAR;
+        KINK_2 = kink2_;
+        JUMP_MULTIPLIER_PER_BLOCK = jumpMultiplierPerYear_ / BLOCKS_PER_YEAR;
+    }
+
+    /**
+     * @notice Calculates the current borrow rate per slot (block)
+     * @param cash The amount of cash in the market
+     * @param borrows The amount of borrows in the market
+     * @param reserves The amount of reserves in the market
+     * @return The borrow rate percentage per slot (block) as a mantissa (scaled by 1e18)
+     */
+    function getBorrowRate(uint256 cash, uint256 borrows, uint256 reserves) external view override returns (uint256) {
+        return _getBorrowRate(cash, borrows, reserves);
+    }
+
+    /**
+     * @notice Calculates the current supply rate per slot (block)
+     * @param cash The amount of cash in the market
+     * @param borrows The amount of borrows in the market
+     * @param reserves The amount of reserves in the market
+     * @param reserveFactorMantissa The current reserve factor for the market
+     * @return The supply rate percentage per slot (block) as a mantissa (scaled by EXP_SCALE)
+     */
+    function getSupplyRate(
+        uint256 cash,
+        uint256 borrows,
+        uint256 reserves,
+        uint256 reserveFactorMantissa
+    ) public view virtual override returns (uint256) {
+        uint256 oneMinusReserveFactor = EXP_SCALE - reserveFactorMantissa;
+        uint256 borrowRate = _getBorrowRate(cash, borrows, reserves);
+        uint256 rateToPool = (borrowRate * oneMinusReserveFactor) / EXP_SCALE;
+        return (utilizationRate(cash, borrows, reserves) * rateToPool) / EXP_SCALE;
+    }
+
+    /**
+     * @notice Calculates the utilization rate of the market: `borrows / (cash + borrows - reserves)`
+     * @param cash The amount of cash in the market
+     * @param borrows The amount of borrows in the market
+     * @param reserves The amount of reserves in the market (currently unused)
+     * @return The utilization rate as a mantissa between [0, EXP_SCALE]
+     */
+    function utilizationRate(uint256 cash, uint256 borrows, uint256 reserves) public pure returns (uint256) {
+        // Utilization rate is 0 when there are no borrows
+        if (borrows == 0) {
+            return 0;
+        }
+
+        return (borrows * EXP_SCALE) / (cash + borrows - reserves);
+    }
+
+    /**
+     * @notice Calculates the current borrow rate per slot (block), with the error code expected by the market
+     * @param cash The amount of cash in the market
+     * @param borrows The amount of borrows in the market
+     * @param reserves The amount of reserves in the market
+     * @return The borrow rate percentage per slot (block) as a mantissa (scaled by EXP_SCALE)
+     */
+    function _getBorrowRate(uint256 cash, uint256 borrows, uint256 reserves) internal view returns (uint256) {
+        int256 util = int256(utilizationRate(cash, borrows, reserves));
```

Comment:
I suppose it would be unlikely, but shouldn't we check here if the cast is safe? what would happen if the utilization rate is greater than the maximum int256?

Reply:
Will it though? I see there are some restrictions on casting, but no runtime checks (https://docs.soliditylang.org/en/latest/080-breaking-changes.html). The issue about runtime checks is still open: https://github.com/ethereum/solidity/issues/11284